pkg/process_manager.go

```diff
@@ -18,46 +26,181 @@ const (
 )
 
 type ProcessStatus struct {
-	// ExitStatus is only valid if `State` is `EXITED`.
-	ExitStatus int32
-	State      ProcessState
+	// ExitCode is only valid if `State` is `EXITED`.
+	ExitCode int
+	State    ProcessState
 }
 
 type processStream struct {
 	// TODO
 }
 
-func (*processStream) Read(p []byte) (n int, err error) {
+func (s *processStream) Read(p []byte) (n int, err error) {
 	panic("TODO")
 }
 
+type process struct {
+	cancel         context.CancelFunc
+	statusLock     sync.RWMutex
+	status         ProcessStatus
+	terminated     atomic.Bool
+	terminatedCond *sync.Cond
+}
+
+func (s *process) markTerminated() {
+	s.terminated.Store(true)
+	s.terminatedCond.Broadcast()
```

Comment:
There's also a particularly tricky to spot deadlock here, because [the docs](https://pkg.go.dev/sync#Cond.Broadcast) are misleading:
> It is allowed but not required for the caller to hold c.L during the call.

...that's only true for the `Cond` itself, i.e. it won't panic regardless of whether you're holding the lock or not. Unfortunately, it's still necessary to protect the condition itself, since `markTerminated()` can trigger a broadcast between `waitTerminated()`'s condition check and `Wait()`, just before that goroutine gets put on the `Cond`'s notify list. It then misses the broadcast and is stuck waiting forever.

Holding the lock can fix this:
```suggestion
	s.terminatedCond.L.Lock()
	s.terminated.Store(true)
	s.terminatedCond.Broadcast()
	s.terminatedCond.L.Unlock()
```

(And since the lock is always held when mutating the variable, you can just use a regular bool flag instead of an atomic)

Reply:
I see. This too should be fixed by switching to channels.